Mathlib/Order/PrimeSeparator.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2024 Sam van Gool. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sam van Gool
+-/
+
+import Mathlib.Order.PrimeIdeal
+import Mathlib.Order.Zorn
+
+
+/-!
+# Separating prime filters and ideals
+
+In a bounded distributive lattice, if $F$ is a filter, $I$ is an ideal, and $F$ and $I$ are
+disjoint, then there exists a prime ideal $J$ containing $I$ with $J$ still disjoint from $F$.
+This theorem is a crucial ingredient to [Stone's][Sto1938] duality for bounded distributive
+lattices. The construction of the separator relies on Zorn's lemma.
+
+## Tags
+
+ideal, filter, prime, distributive lattice
+
+## References
+
+* [M. H. Stone, Topological representations of distributive lattices and Brouwerian logics
+(1938)][Sto1938]
+-/
+
+
+universe u
+variable {α : Type*}
+
+open Order Ideal Set
+
+variable [DistribLattice α] [BoundedOrder α]
+
+variable {F : PFilter α} {I : Ideal α}
+
+namespace DistribLattice
+
+
+lemma mem_ideal_sup_principal (a b : α) (J : Ideal α) : b ∈ J ⊔ principal a ↔ ∃ j ∈ J, b ≤ j ⊔ a :=
+  ⟨fun ⟨j, ⟨jJ, _, ha', bja'⟩⟩ => ⟨j, jJ, le_trans bja' (sup_le_sup_left ha' j)⟩,
+    fun ⟨j, hj, hbja⟩ => ⟨j, hj, a, le_refl a, hbja⟩⟩
+
+theorem prime_ideal_of_disjoint_filter_ideal (hFI : Disjoint (F : Set α) (I : Set α)) :
+    ∃ J : Ideal α, (IsPrime J) ∧ I ≤ J ∧ Disjoint (F : Set α) J := by
+
+  -- Let S be the set of proper ideals containing I and disjoint from F
+  set S : Set (Set α) := { J : Set α | IsIdeal J ∧ I ≤ J ∧ Disjoint (F : Set α) J }
```

Comment:
couldn't you define S as a `Set (Ideal α)`?

Reply:
I tried this quite seriously several times and would have indeed liked the definition to look like that. Unfortunately I don't see how to do it at the moment, for the following reason. In the proof of `chainub` a bit below, I need at some point to take a union of a chain of ideals in `S`. Mathlib has a complete lattice instance on `Ideal α`, so I can take the `sSup` instead of the `sUnion`, but the supremum of a set of ideals is not equal to a union in general. It is only a union in this case, because we are taking the supremum of a chain. So, if one wanted to make this work, it would require first proving (somewhere else) that `sSup` of a chain of ideals is a union. I'm not sure this is worthwhile.

Reply:
For standard rings, sSup of a chain of ideals is a union, and this is very useful for proving the existence of maximal ideals.

Reply:
I would personally be in favor of implementing that for lattice ideals as well but I would imagine it is a separate PR.